### PR TITLE
layers: support BigData

### PR DESCRIPTION
**Pull request by giuseppe**
similarly to containers and images, add support for storing big data
also for layers, so that it is possible to store arbitrary data when
it is not feasible to embed it in the layers JSON metadata file.

Subset of: https://github.com/containers/storage/pull/775
    
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
@nalind @rhatdan PTAL

**Comment by rhatdan** [comment]
LGTM
@vrothberg @mtrmac @TomSweeneyRedHat PTAL

File: cmd/containers-storage/layer.go
Comment (TomSweeneyRedHat) [inline cmd/containers-storage/layer.go:166]: nit
```suggestion
			usage:       "Get data that is attached to a layer",
```

File: cmd/containers-storage/layer.go
Comment (TomSweeneyRedHat) [inline cmd/containers-storage/layer.go:176]: nit
```suggestion
			usage:       "Set data that is attached to a layer",
```

File: docs/containers-storage-get-layer-data.md
Comment (TomSweeneyRedHat) [inline docs/containers-storage-get-layer-data.md:1]: I'm a bit confused over exactly what data is being gathered.  Title of the PR says "BigData", functions about Get/Set LayerBigData, help above "data attached to a layer", and man page says "lookaside" data.

Can we be consistent through out?  I'd prefer "lookaside data" if that holds.

File: cmd/containers-storage/layer.go
Comment (giuseppe) [reply cmd/containers-storage/layer.go]: thanks, fixed

File: cmd/containers-storage/layer.go
Comment (giuseppe) [reply cmd/containers-storage/layer.go]: thanks, fixed!

File: docs/containers-storage-get-layer-data.md
Comment (giuseppe) [reply docs/containers-storage-get-layer-data.md:1]: it can be any data you want to record for a layer.  In particular, it is going to record the zstd:chunked manifest in https://github.com/containers/storage/pull/775

Comment (TomSweeneyRedHat) [reply docs/containers-storage-get-layer-data.md:1]: OK, but my comment still holds.  It appears we're calling it three or four different things inside the code and on the man pages.  Can we change the variable names/function calls to refer to lookaside data?

**Review by mtrmac — COMMENTED** [review]
A literally one-minute impression, and design concerns like this are definitely decided by @nalind , not me:

This can make sense as an internal implementation mechanism; it’s not obvious to me that this _must_ be exposed as a raw data to external callers. If c/image is only going to consume that data through a zstd:chunked - specific API, only that zstd:chunked - specific API needs to be public.

Making raw data public like this
- effectively freezes the format of the raw data, slowing down future improvements.
- invites other uses of raw data for unexpected purposes. That can be very useful and allow for unexpected innovations, OTOH it also further cements down the concept of individual layers and makes it even harder to significantly restructure (e.g. to a file-based approach like OSTree or the chunked format).

Again, definitely up to @nalind, just airing my first impression.

File: docs/containers-storage-get-layer-data.md
Comment (giuseppe) [reply docs/containers-storage-get-layer-data.md:1]: BigData is the name already used for containers and images.  I think a different name can be confusing.

Should I change the comments?

Comment (TomSweeneyRedHat) [reply docs/containers-storage-get-layer-data.md:1]: @giuseppe I'm fine with whatever name you use, my main issue was you used 3 or more names for it.  So if you did this:

LayerBigData  -> keep as is
"data attached to a layer" -> "BigData attached to a layer"
"lookaside" data -> BigData  or "the layer's BigData"

I'd be good with that.

File: cmd/containers-storage/container.go
Comment (mtrmac) [inline cmd/containers-storage/container.go:15]: I’d expect this to go to `…/layer.go`.

File: docs/containers-storage-get-layer-data.md
Comment (mtrmac) [inline docs/containers-storage-get-layer-data.md:1]:
```suggestion
## containers-storage-get-layer-data 1 "December 2020"
```

File: docs/containers-storage-set-layer-data.md
Comment (mtrmac) [inline docs/containers-storage-set-layer-data.md:1]:
```suggestion
## containers-storage-set-layer-data 1 "December 2020"
```

File: layers.go
Comment (mtrmac) [inline layers.go:292]: `copyStringSlice`?

File: layers.go
Comment (mtrmac) [inline layers.go:128]: Values of this are AFAICS never set.

File: layers.go
Comment (mtrmac) [inline layers.go:996]: How big is “big”? Does it make sense for the API to read/write using `[]byte`, or should that be a streaming interface?

File: layers.go
Comment (mtrmac) [inline layers.go:128]: Just reading the API documentation, I would expect `BigDataSizes` to be included in `ImageSize`. (Otherwise it’s not even clear to me there needs to be a separate faster-to-access record of the sizes.)

**Review by mtrmac — COMMENTED** [review]
A few more notes. I didn’t read all of the code carefully, and I completely defer to actual c/storage maintainers for all the design decisions.

File: layers.go
Comment (mtrmac) [reply layers.go:996]: For a bit more context, this is probably not a concern for fairly fixed things like manifests (in image “big data”). OTOH there _has_ been a report of Raspberry Pis running out of memory, probably when loading tar-split metadata into RAM.

File: docs/containers-storage-get-layer-data.md
Comment (giuseppe) [reply docs/containers-storage-get-layer-data.md:1]: I am still unsure.  LayerBigData is about the API while "data attached to a layer" and "lookaside data" are used in the documentation and code comments.  Using `BigData` to explain what the Layer BigData is seems confusing:

e.g. converting:
`containers-storage get-layer-data - Retrieve lookaside data for a layer`
to:
`containers-storage get-layer-data - Retrieve bigdata for a layer`

does not seem like an improvement

File: layers.go
Comment (giuseppe) [reply layers.go:128]: thanks for spotting it.  I am going to push a fixed version

Comment (giuseppe) [reply layers.go:128]: I've only added it because it was present in the image API.  I am fine with dropping it

File: docs/containers-storage-get-layer-data.md
Comment (TomSweeneyRedHat) [reply docs/containers-storage-get-layer-data.md]: I'll let you make the final decision @giuseppe .  Not a real biggy, I just think it's a bit confusing as is.  I'll trust your judgment on the final naming.

Comment (rhatdan) [reply docs/containers-storage-get-layer-data.md]: I prever
 containers-storage get-layer-data - Retrieve lookaside data for a layer

File: layers.go
Comment (rhatdan) [reply layers.go:128]: Let's drop it.

Comment (mtrmac) [reply layers.go:128]: Why _shouldn’t_ the size be included in `ImageSize`?

Comment (giuseppe) [reply layers.go]: > Let's drop it.

dropped and pushed a new version.

**Comment by giuseppe** [comment]
is there anything more left to address?

**Comment by rhatdan** [comment]
@nalind @mtrmac PTAL

**Comment by mtrmac** [comment]
See the 2 points from my previous review.

File: layers.go
Comment (giuseppe) [reply layers.go:990]: "big" is not really that big, the generated manifest file for Fedora is around 3M

File: layers.go
Comment (giuseppe) [reply layers.go]: > Why _shouldn’t_ the size be included in `ImageSize`?

it is not used at the moment, we can add it if there is any user of it

**Comment by giuseppe** [comment]
> See the 2 points from my previous review.

just replied to them

File: layers.go
Comment (mtrmac) [reply layers.go:990]: We’ve had reports like https://github.com/containers/image/issues/495 (“hundreds of gigabytes”); and on the other side, users on Raspberry Pi with IIRC 256 MB of RAM.

I’m not yet saying that this is not good enough, right now I’m just pointing out that Fedora is _by far_ not the worst case of how far this should scale.

File: layers.go
Comment (mtrmac) [reply layers.go]: The Podman repo suggests that `store.ImageSize` is called for “disk usage statistics” (`df`?), “inspect”, “prune” and “tree”; I didn’t trace the full call stack, it may be exposed via other features or it might be dead code.

OTOH it’s true that c/image’s `types.Image.Size()` implementation does its own calculation which does not currently benefit from extending `store.ImageSize`.

**Comment by giuseppe** [comment]
@mtrmac @rhatdan anything more holding this PR?

**Comment by rhatdan** [comment]
Not from me. Just waiting for @mtrmac comment.

**Comment by mtrmac** [comment]
\*shrug\* I don’t have anything to add WRT the implementation; I have some concerns but they are not quite blocking.

As for the more general design review, I know very little about c/storage , that would be up to @nalind I think.

**Comment by giuseppe** [comment]
how to move forward?  This is just a subset of the changes I'd like to get merged for zstd:chunked support

**Comment by rhatdan** [comment]
@vrothberg @nalind PTAL so we can get this merged.

**Comment by vrothberg** [comment]
Added to my todo list.  I won't make it today but promise to have a look on Monday.

Apologies for not having looked into it until now!

**Review by rhatdan — APPROVED** [review]
Setting approved

File: layers.go
Comment (vrothberg) [reply layers.go:990]: I echo @mtrmac's concern.  Returning an `io.ReadCloser` seems like a safe(r) bet.

File: layers.go
Comment (vrothberg) [inline layers.go:1001]: Same here, we could turn `data` into an `io.Reader`?

File: layers.go
Comment (vrothberg) [inline layers.go:996]: Can we add "big data" to the error message?  That'll help to put a finger on the code in case of errors.

Comment (vrothberg) [reply layers.go:996]: There are other cases below

File: layers.go
Comment (giuseppe) [reply layers.go:1001]: I am fine with it, just it will be different than all the others `BigData()/SetBigData()` functions we already have for containers and images

Comment (giuseppe) [reply layers.go]: changed to use `io.Reader and io.ReadCloser`

File: layers.go
Comment (giuseppe) [reply layers.go:996]: fixed

**Review by vrothberg — COMMENTED** [review]
LGTM

File: layers.go
Comment (mtrmac) [inline layers.go:998]: Note, to be explicit: actual reads happen outside of the layer lock. AFAICS this is safe because removing the file does not truncate the underlying inode, and writes happen via `NewAtomicFileWriter`, i.e. don’t overwrite the original.

(Perhaps add comment to the `NewAtomicFileWriter` call below to note that the read implementation relies on it?)

Comment (giuseppe) [reply layers.go:998]: thanks.

Added a comment

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**
